presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/Extensions.kt

```diff
@@ -0,0 +1,241 @@
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+typealias OnChange<T> = (T) -> Unit
+
+/**
+ * Convenience method for creating a [DistinctObservableProperty].
+ * @param onChange an optional block that will be executed when the property has changed.
+ * @param onChangeOnInit if true, the provided [onChange] block will be executed with the [initialValue]. The default is true.
+ *
+ */
+fun <T> distinctObservable(
+        initialValue: T,
+        onChange: OnChange<T>? = null,
+        onChangeOnInit: Boolean = true):
+        ReadWriteProperty<Any?, T> = object : DistinctObservableProperty<T>(initialValue) {
+
+    init {
+        if (onChangeOnInit)
+            onChange?.invoke(initialValue)
+    }
+
+    override fun afterChange(property: KProperty<*>, oldValue: T, newValue: T) {
+        onChange?.invoke(newValue)
+    }
+}
+
+/**
+ * Creates a [BindableProperty] for properties of classes that extend [BaseObservable].
+ *
+ * The property must be annotated with [Bindable].
+ *
+ * Usage with all parameters:
+ * ```
+ * @get:Bindable
+ * var firstName: String by bindable("", BR.firstName)
+ * ```
+ *
+ * Usage without explicit resource ID and type inference:
+ * ```
+ * @get:Bindable
+ * var firstName by bindable("")
+ *
+ * ```
+ * @param bindableResourceId the id from the generated `BR` class. If omitted, the [PropertyCacher] must be initialised prior to usage.
+ */
+fun <T> BaseObservable.bindable(
+        initialValue: T,
+        bindableResourceId: Int = USE_PROPERTY_CACHER)
+        : BindableProperty<T, NotifiableObservable> {
+
+    return object : BindableProperty<T, NotifiableObservable>(initialValue, bindableResourceId) {
+        override fun notify(bindable: NotifiableObservable, propertyId: Int, newValue: T) {
+            bindable.notifyPropertyChanged(propertyId)
+        }
+    }
+}
+
+/**
+ * Creates a [BindableProperty] for properties of classes that extend [BaseObservable].
+ *
+ * The property must be annotated with [Bindable].
+ *
+ * Usage with all parameters:
+ * ```
+ * @get:Bindable
+ * var firstName: String by bindable("", BR.firstName) {
+ *      // do something
+ * }
+ * ```
+ *
+ * Usage without explicit resource ID and type inference:
+ * ```
+ * @get:Bindable
+ * var firstName by bindable("") {
+ *      // do something
+ * }
+ *
+ * ```
+ * @param bindableResourceId the id from the generated `BR` class. If omitted, the [PropertyCacher] must be initialised prior to usage.
+ * @param onChange an optional block that will be executed after the property has changed, executed also on the initial value.
+ * @param onChangeOnInit if true, the provided [onChange] block will be executed with the [initialValue].
+ */
+inline fun <T> BaseObservable.bindable(
+        initialValue: T,
+        bindableResourceId: Int = USE_PROPERTY_CACHER,
+        crossinline onChange: OnChange<T>,
```

Comment:
Ideally this function should be the only implementation (where the one above disappears).

Due to a limitation with current Kotlin, we cannot provide defaults for inlined lambda expressions, the following yields an error:
`onChange: OnChange<T> = {}  // construction is not yet supported in inline functions`